File: stage-10-shop-admin/django/shop/store.py
import hashlib
import json
import secrets
from pathlib import Path

from django.conf import settings


def _read(path: Path, fallback):
    if not path.is_file():
        return fallback
    with path.open(encoding="utf-8") as handle:
        data = json.load(handle)
    return data if isinstance(data, list) or isinstance(data, dict) else fallback


def load_products():
    return _read(settings.PRODUCTS_FILE, [])


def load_store():
    store = _read(settings.STORE_FILE, {})
    store.setdefault("users", [])
    store.setdefault("orders", [])
    return store


def save_store(store):
    path = Path(settings.STORE_FILE)
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(store, indent=2), encoding="utf-8")


def hash_password(password: str) -> str:
    salt = secrets.token_hex(16)
    digest = hashlib.pbkdf2_hmac("sha256", password.encode(), salt.encode(), 120000)
    return f"pbkdf2${salt}${digest.hex()}"


def check_password(password: str, stored: str) -> bool:
    try:
        algo, salt, digest = stored.split("$", 2)
    except ValueError:
        return False
    if algo != "pbkdf2":
        return False
    check = hashlib.pbkdf2_hmac("sha256", password.encode(), salt.encode(), 120000)
    return secrets.compare_digest(check.hex(), digest)


def public_user(user):
    return {"id": user["id"], "name": user["name"], "email": user["email"]}


def find_product(product_id: str):
    for product in load_products():
        if product.get("id") == product_id:
            return product
    return None


def find_user_by_email(store, email: str):
    email = email.lower()
    for user in store["users"]:
        if user["email"].lower() == email:
            return user
    return None


def find_user_by_id(store, user_id):
    for user in store["users"]:
        if user["id"] == user_id:
            return user
    return None

File: stage-09-database/nodejs/store.js
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");

const productsFile = path.join(__dirname, "..", "data", "products.json");
const storeFile = path.join(__dirname, "data", "store.json");

function readJson(file, fallback) {
  if (!fs.existsSync(file)) return fallback;
  try {
    return JSON.parse(fs.readFileSync(file, "utf8"));
  } catch (error) {
    return fallback;
  }
}

function loadProducts() {
  const data = readJson(productsFile, []);
  return Array.isArray(data) ? data : [];
}

function loadStore() {
  const store = readJson(storeFile, {});
  if (!Array.isArray(store.users)) store.users = [];
  if (!Array.isArray(store.orders)) store.orders = [];
  return store;
}

function saveStore(store) {
  fs.mkdirSync(path.dirname(storeFile), { recursive: true });
  fs.writeFileSync(storeFile, JSON.stringify(store, null, 2));
}

function hashPassword(password) {
  const salt = crypto.randomBytes(16).toString("hex");
  const digest = crypto.pbkdf2Sync(password, salt, 120000, 32, "sha256").toString("hex");
  return `pbkdf2$${salt}$${digest}`;
}

function checkPassword(password, stored) {
  try {
    const parts = String(stored).split("$");
    if (parts.length !== 3 || parts[0] !== "pbkdf2") return false;
    const digest = crypto.pbkdf2Sync(password, parts[1], 120000, 32, "sha256");
    const actual = Buffer.from(parts[2], "hex");
    if (digest.length !== actual.length) return false;
    return crypto.timingSafeEqual(digest, actual);
  } catch (error) {
    return false;
  }
}

function publicUser(user) {
  return { id: user.id, name: user.name, email: user.email };
}

function findProduct(id) {
  return loadProducts().find((product) => product.id === id) || null;
}

function findUserByEmail(store, email) {
  const needle = email.toLowerCase();
  return store.users.find((user) => user.email.toLowerCase() === needle) || null;
}

function findUserById(store, id) {
  return store.users.find((user) => user.id === id) || null;
}

module.exports = {
  loadProducts,
  loadStore,
  saveStore,
  hashPassword,
  checkPassword,
  publicUser,
  findProduct,
  findUserByEmail,
  findUserById,
};
